source/configuration_manual/howto/director_with_lua.rst

```diff
@@ -0,0 +1,115 @@
+==============================
+
+
+.. note:: Lua director is not supported for Dovecot Pro.
+
+Small scale deployments using NFS or other shared storage need a way to ensure users
+end up in same backend. For this purpose, a simple Lua script can be used.
+
+The script can be run on a proxy directly, or one can have one dedicated proxy node acting
+as a Lua director node running the script.
+
+Prerequisites
+-------------
+
+This script can use any SQL driver supported by Lua DBI.
+If you want to use SQLite, or some other file based database, you can only run this on single node.
+
+You will need `Lua DBI <https://github.com/mwild1/luadbi>`_, and a database driver for it.
+For the shard version, you will need to install crc32 library.
+
+Schema
+------
+
+Example SQL schema for the direct user to backend mapping.
+
+.. code:: sql
+
+  CREATE TABLE backends (
+     id int not null primary key,
+     hostname varchar(255) not null unique,
+     hostip varchar(255), 
+     state smallint not null default 0
+  );
+  
+  CREATE TABLE user_backend (
+     backend_id int not null,
+     user varchar(255),
+     primary key(backend_id, user)
+  );
+
+And schema for the sharded version.
+
+.. code:: sql
+
+  CREATE TABLE backends (
+     id int not null primary key,
+     hostname varchar(255) not null unique,
+     hostip varchar(255), 
+     state smallint not null default 0
+  );
+  
+  CREATE TABLE user_backend (
+     backend_id int not null,
+     user_hash int not null,
+     primary key(backend_id, user_hash)
+  );
+
+
+Lua script
+----------
+
+The Lua script comes in two variants, first one simply does username <-> hostname mapping,
+and can be found at https://github.com/dovecot/tools/blob/main/director.lua.
+The second does sharding and can be found at https://github.com/dovecot/tools/blob/main/director-shard.lua.
+
+In database, the backend\_ip field can be left empty when it's not needed.
+It can be used when your backend names cannot be resolved.
+
+Configuration
+-------------
+
+There are few ways you can use this script.
+First, and the most simple, is drop-in replacement for director with no authentication.
+
+.. code::
+
+  passdb {
+    driver = lua
+    args = file=/etc/dovecot/director.lua nopassword
+    # or password=masterpass
+  }
+
+If you wish to do authentication, you can do
+
+.. code::
+
+  passdb {
+    driver = pam
+    result_success = continue-ok
+  }
+
+  passdb {
+    skip = unauthenticated
+    driver = lua
+    args = file=/etc/dovecot/director.lua noauthenticate
+  }
+
+
+If you are using MySQL or PostgreSQL, you can also install this directly on your proxy node(s),
+and skip having a centralized director node.
+
+Operations
+----------
+
+There are no built-in tools in Dovecot to manage the state in the database, you need to
+build your own tooling.
+
+To move a user, update the database record and then kick the user from the backend.
+This should be done rather quickly if the user is active.
```

Comment:
This reads a bit handwavy and unspecific.

I would focus on explaining the necessity to be quick. Does the user need to be synchronized? Does the operation need to be atomic? Does the user need to be locked in some way or another. The next paragraph could be integrated into this one to present one option.

Otherwise you can be sure it will lead to problems and at the very least repetitive questions on multiple channels.

Reply:
I think I'll just drop these.